[Chua Feng Yuan] iP

srfl0 opened the pull request.

Commit:
Increment level-0

Commit:
Increment level-0

Commit:
Add interfacing for commands

Add echo() command to chatbot

Commit:
Add functionalities to store user inputs and repeat them upon user request

Add comments to clarify code

Commit:
Add Task class to OOP-ify tasks for simpler processing of list
Add functionalities to manipulate Task class for use in program

Commit:
Add behaviour to chatbot for attempting to modify state of a task to identical current state

szeyingg commented on src/main/java/Ixo.java:
```diff
@@ -0,0 +1,221 @@
+
+    }
+
+    public static void add() {
```
method name could be more specific

szeyingg commented on src/main/java/Ixo.java:
```diff
@@ -0,0 +1,221 @@
+
+        String[] store = new String[50]; //limit list to store only 50 items
+        int storeIndex = 0;
+        String newline;
```
variable name should be "newLine" to follow coding standards

szeyingg commented on src/main/java/Ixo.java:
```diff
@@ -0,0 +1,221 @@
+            line = newline.split(" ");
+
+            switch (line[0]) {
+            case "bye":
+                return;
+
+            case "list":
```
remove extra line in between cases for to follow standard layout

szeyingg commented on src/main/java/Ixo.java:
```diff
@@ -0,0 +1,221 @@
+            storeIndex++;
+            System.out.println();
+            System.out.println(separator);
+        }
+
+    }
```
can remove extra space between brackets

szeyingg commented on src/main/java/Ixo.java:
```diff
@@ -0,0 +1,221 @@
+
+    }
+
+    public static void tasklist() {
```
function name should be properly capitalised, change to "taskList"

szeyingg commented on src/main/java/Ixo.java:
```diff
@@ -0,0 +1,221 @@
+            storeIndex++;
+            System.out.println();
+            System.out.println(separator);
+        }
+
+    }
```
extra space

szeyingg reviewed (commented):
code mostly follows coding standards, but could be more mindful about redundant spacings between lines of code as it sometimes affects readability.

Lukapeng77 commented on src/main/java/Ixo.java:
```diff
@@ -0,0 +1,221 @@
+        do {
+            switch (cmd.toLowerCase()) {
+            case "echo":
+            case "1":
+                echo();
+                break;
+
+            case "add":
+            case "2":
+                add();
+                break;
+
+            case "tasking":
+            case "3":
+                tasklist();
+                break;
```
For the case types, instead of using the magic numbers "1", "2", "3", can just keep like the "add", "echo", 'list" which can be more meaningful.

Lukapeng77 commented on src/main/java/Ixo.java:
```diff
@@ -0,0 +1,221 @@
+                }
+                continue;
+
+            case "mark":
+                storeIndexCheck = Integer.parseInt(line[1])-1;
+                Task itemToMark = store[storeIndexCheck];
+                if (itemToMark.isDone) {
+                    System.out.println("You have already done this task.");
+                }
+                else {
+                    System.out.println("Well done! I've marked this task as done");
+                    itemToMark.isDone = true;
+                    System.out.println("[" + itemToMark.getStatusIcon() + "] " + itemToMark.description);
+                }
+                System.out.println(separator);
+                continue;
+
+            case "unmark":
+                storeIndexCheck = Integer.parseInt(line[1])-1;
+                Task itemToUnmark = store[storeIndexCheck];
+                if (!itemToUnmark.isDone) {
+                    System.out.println("This task was not done yet.");
+                }
+                else {
+                    System.out.println("Okay, I'll mark this task as not done");
+                    itemToUnmark.isDone = false;
+                    System.out.println("[" + itemToUnmark.getStatusIcon() + "] " + itemToUnmark.description);
+                }
+                System.out.println(separator);
+                continue;
+            }
```
Consider resue the same part of code snippets for the mark and unmark parts to avoid unnecessary redundancy and make use of the codes.

Lukapeng77 commented on src/main/java/Ixo.java:
```diff
@@ -0,0 +1,221 @@
+            System.out.print("added: ");
+            for (String word : line) {
+                System.out.print(word);
+                space = (word.equals(line[line.length - 1])) ? "" : " ";
```
It's better to create a new variable to store the line.length -1 value and better explain the variable meaning, avoid using magic numbers.

Lukapeng77 commented on src/main/java/Ixo.java:
```diff
@@ -0,0 +1,221 @@
+
+        String space; //variable to determine if spacing is needed for purposes of echoing
+
+        while (true) {
```
For check the continue condition, can create a new variable to store and explain the meaning of the variable.
Eg: Boolean isExist = false; To check for the continue condition.

Commit:
Remove add() and echo() features
Add functionalities for adding todo, deadlines and events to tasklist
Abstract mark and check features from task list as operations within a single-branch

To do: input filtering through try-throw-catch

Commit:
Modify string-parsing of cmd in taskList() to always be for lowercase

Commit:
Add exception handling

Commit:
Comment on catch statement reasoning

Commit:
Merge branch 'Branch-Level-5'
Merge from branch-level-5

sevenseasofbri commented on src/main/java/Ixo.java:
```diff
@@ -0,0 +1,151 @@
+
+    public final static String SEPARATOR = "____________________________________________________";
+
+    public static void menu() {
+        System.out.println("1. Task List");
+    }
```
Function names should ideally be verbs. So a name like `showMenu` or `printMenu` would be more apt.

sevenseasofbri commented on src/main/java/Ixo.java:
```diff
@@ -0,0 +1,151 @@
+        System.out.println("1. Task List");
+    }
+
+    public static void taskList() {
```
Similarly for this, function names should be verbs. Also, the function name does not appear to match what is happening inside it. Additionally, thif function is >30LOC. Consider breaking it up into multiple smaller functions.

sevenseasofbri commented on src/main/java/Ixo.java:
```diff
@@ -0,0 +1,151 @@
+    public static void taskList() {
+
+        Scanner lineScan;
+        Task[] taskStore = new Task[100]; //limit list to store only 100 items
```
Try and avoid the usage of magic literals by storing them in variables or constants. Something like `MAX_TASKS` or something would be better.

sevenseasofbri commented on src/main/java/Ixo.java:
```diff
@@ -0,0 +1,151 @@
+
+            case "mark":
+            case "unmark":
+                try {
+                    taskToMark = Integer.parseInt(inputLine.split(" ")[1]) - 1; //-1 to match up user view to actual index access
+                    if (taskToMark >= storeIndex || taskToMark < 0) {
+                        System.out.println("that task does not exist");
+                        System.out.println(SEPARATOR);
+                    }
+                    boolean isMark = cmd.equals("mark");
+                    if (isMark) {
+                        System.out.println("Well done! task " + taskStore[taskToMark] + " has been marked.");
+                    } else {
+                        System.out.println("The task " + taskStore[taskToMark] + " has been unmarked.");
+                    }
+                    taskStore[taskToMark].isDone = isMark;
+                } catch (NumberFormatException e) { //for non-number input
+                    System.out.println("You did not enter a number");
+                } catch (ArrayIndexOutOfBoundsException e) { //for out of bounds index entered
+                    System.out.println("You did not enter a valid corresponding number");
+                }
```
This is deep nesting and can borderlines arrowhead code. Refactor the internal indents to another function.

sevenseasofbri commented on src/main/java/Ixo.java:
```diff
@@ -0,0 +1,151 @@
+    }
+
+
+    public static void command() {
```
Again, the name of this function is not apt. It needs to be a verb. 
```suggestion
    public static void parseCommand() {
```

sevenseasofbri reviewed (commented):
It appears as though level 6 and 7 have not been completed yet? Please complete it soon. Do look into the java coding standard, because the naming for the functions needs to be improved.

Commit:
Refresh Ixo for implementation using ArrayList instead of array
Add feature to delete tasks
Add necessary basic exceptions

Commit:
Merge branch 'branch-level-6'

Commit:
Add functionality to save current list of tasks into a text file and load upon starting up program

Commit:
Collate all files under Ixo package

Commit:
Add some progress into doing OOP-ifying

Commit:
Refactor and clean up code for OOP-ification

Commit:
Amend missing rename for run method in Ui (now called begin)

Commit:
Add find functionality for Ixo

Commit:
Merge branch 'branch-level-9'

Commit:
Update .gitignore to avoid committing data text files

Commit:
Add JavaDoc descriptions
Add icon design on welcome

Commit:
Create user guide
Adjust JavaDoc annotations

Commit:
Rectify issues arising from usage of Java 23

Commit:
Add functionality to sanitise and control user inputs against injection-type crashes

Commit:
Modify welcome message
